ewatercycle/models/pcrglobwb.py

```diff
@@ -11,112 +11,195 @@
+                    kwargs['routing_method'])
 
-        for section, options in kwargs.items():
-            for option, value in options.items():
-                cfg.set(section, option, value)
+        if "dynamic_flood_plain" in kwargs:
+            cfg.set("routingOptions", "dynamicFloodPlain",
+                    kwargs["dynamic_flood_plain"])
 
+        if "max_spinups_in_years" in kwargs:
+            cfg.set("globalOptions", "maxSpinUpsInYears",
+                    str(kwargs["max_spinups_in_years"]))
+
+    def _export_config(self) -> PathLike:
         new_cfg_file = Path(self.work_dir) / "pcrglobwb_ewatercycle.ini"
         with new_cfg_file.open("w") as filename:
-            cfg.write(filename)
+            self.config.write(filename)
 
-        self.cfg_file = new_cfg_file.resolve()
-        print(f"Created config file {self.cfg_file} with inputDir "
-              f"{full_input_path} and outputDir {self.work_dir}.")
+        self.cfg_file = new_cfg_file.expanduser().resolve()
+        return self.cfg_file
 
-    def _start_container(self,
-                         input_dir: PathLike,
-                         additional_input_dirs: Iterable[PathLike] = []):
-        input_dirs = [input_dir] + list(additional_input_dirs)
+    def _start_container(self):
+        additional_input_dirs = [
+            str(self.parameter_set.input_dir),
+            str(self.forcing.directory)
+        ] + self._additional_input_dirs
 
         if CFG["container_engine"] == "docker":
             self.bmi = BmiClientDocker(
-                image=CFG["pcrglobwb.docker_image"],
+                image=self.docker_image,
                 image_port=55555,
                 work_dir=str(self.work_dir),
-                input_dirs=[str(input_dir) for input_dir in input_dirs],
+                input_dirs=additional_input_dirs,
                 timeout=10,
             )
         elif CFG["container_engine"] == "singularity":
-            image = CFG["pcrglobwb.singularity_image"]
-
-            message = f"No singularity image found at {image}"
-            assert Path(image).exists(), message
-
-            self.bmi = BmiClientSingularity(
-                image=image,
-                work_dir=str(self.work_dir),
-                input_dirs=[str(input_path) for input_path in input_dirs],
-                timeout=10,
-            )
+            try:
+                self.bmi = BmiClientSingularity(
+                    image=f"docker://{self.docker_image}",
+                    work_dir=str(self.work_dir),
+                    input_dirs=additional_input_dirs,
+                    timeout=15,
+                )
+            except FutureTimeoutError:
+                raise ValueError(
+                    "Couldn't spawn the singularity container within allocated"
+                    " time limit (15 seconds). You may try building it with "
+                    f"`!singularity run docker://{self.docker_image}` and try "
```

Comment:
When I got this exception in the example notebook, the building of the Singularity image kept going (as evident from the Jupyter log). So I let that complete and then ran setup again.

The `singularity run docker://ewatercycle/pcrg-grpc4bmi:setters` will run forever as it starts a BMI server. My suggestion would be to run `singularity exec docker://ewatercycle/pcrg-grpc4bmi:setters run-bmi-server -h` as it will show some help and then immediately exit.
Also having a `!` assumes that the error was thrown in a notebook. I don't think its wise to presume that and would remove the `!`.

The last sentence assumes the system admin has diverted the SINGULARITY_CACHEDIR to a shared location instead of the default `~/.singularity`. This is not possible according to https://singularity.hpcng.org/user-docs/3.7/build_env.html#sec-cache

Reply:
Aii, that's a good point. Maybe we cannot use the `docker://` syntax after all then, if we want the admin to manage the containers on the system. 

For now I've wrapped the whole container startup in a try/except block and adapted the message.

Reply:
https://github.com/eWaterCycle/ewatercycle/issues/119